README.md
# Notice

I've rewritten this tool using Swift ([`sendkeys`](https://github.com/socsieng/sendkeys)) and can be installed using
[homebrew](https://brew.sh/):

```sh
brew install socsieng/tap/sendkeys
```

It includes expanded support for mouse commands.

I am no longer maintaining this library and encourage you to use the Swift version instead.

# SendKeys for macOS

This is a CLI tool to send keys to a given app to simulate typing at a given speed.

This tool was originally created to make it easier to enter keystrokes for a smooth screen recording of code being typed
in.

_Includes experimental support for mouse operations._

# Installation

```sh
npm install sendkeys-macos --global
```

![npm install animation example](https://github.com/socsieng/sendkeys-macos/raw/master/docs/sendkeys.gif)

## CLI usage

Basic usage:

```sh
sendkeys -a "Notes" -c "Hello<p:1> world<c:left:option,shift><c:i:command>"
```

![hello world example](https://github.com/socsieng/sendkeys-macos/raw/master/docs/example1.gif)

Providing input from a file:

```sh
sendkeys -a "Visual Studio Code" -f example.txt
```

Receiving input from `stdio`:

```sh
cat example.txt | sendkeys -a "Notes"
```

Refer to the help command for more options:

```sh
sendkeys --help
```

## Node usage

Programatic usage:

```js
const sendKeys = require('sendkeys-macos');

sendKeys('Notes', 'hello<c:a:command><c:c:command><c:right> <c:v:command>', { delay: 0.1, initialDelay: 1 });
```

![hello hello example](https://github.com/socsieng/sendkeys-macos/raw/master/docs/example2.gif)

## Markup instructions

Basic markup is supported to control the pause between keystrokes and to apply additional keystroke combinations.

### Inserting a pause

A pause can be inserted with the `<p:seconds>` sequence.

By default, the delay between keystrokes is set with the `-d` or `--delay` option where the default is 0.1 seconds. A
one-off pause of 1 second can be applied between characters by inserting `<p:1>`.

`<P:seconds>` (note upper case `P`) can be used to modify the default delay between subsequent keystrokes.

### Special key combinations

Special key combinations including non-printable character sequences can be applied using the `<c:key>` or
`<c:key:modifiers>` markup.

`key` can include any printable character or, one of the following key names: `f1`, `f2`, `f3`, `f4`, `f5`, `f6`, `f7`,
`f8`, `f9`, `f10`, `f11`, `f12`, `esc`, `return`, `enter`, `delete`, `space`, `tab`, `up`, `down`, `left`, `right`,
`home`, `end`, `pgup`, and `pgdown`.

`modifiers` is an optional list of comma separated values that can include `command`, `shift`, `control`, and `option`.

Example key combinations:

- `tab`: `<c:tab>`
- `command` + `a`: `<c:a:command>`
- `option` + `shift` + `left arrow`: `<c:left:option,shift>`

### Mouse commands

Experimental support for mouse commands, including:

- Mouse move: `<m:x1,y1,x2,y2:duration_in_seconds>`
- Mouse click: `<m:button:number_of_clicks>`
  - `button` supported values `left`, `center`, `right`.
  - `number_of_clicks` defaults to `1`
- Mouse drag: `<d:x1,y1,x2,y2:duration_in_seconds:button>`
  - `button` supported values `left`, `center`, `right`. Defaults to `left`.

### Continuation

A continuation can be used to ignore the next character. This is useful to help with formatting a long sequence of
character and inserting a new line for authoring purposes.

Insert a continuation using the character sequence `<\>`. The character following the sequence will be skipped over.

## Prerequisites

This script only works on macOS as it has a dependency on the macOS Automator application.

When running from the terminal, ensure that the terminal has permissions to send keystrokes. This can be done by
navigating to System Preferences > Security & Privacy > Privacy > Accessibility and adding your terminal application
there.


CHANGELOG.md
# Changelog

## [1.3.0](https://www.github.com/socsieng/sendkeys-macos/compare/v1.2.2...v1.3.0) (2020-12-31)


### Features

* output current mouse position when no arguments are specified ([4890714](https://www.github.com/socsieng/sendkeys-macos/commit/48907140d0eeb52bb4da6504f7b5b06f69bcb438))

### [1.2.2](https://www.github.com/socsieng/sendkeys-macos/compare/v1.2.1...v1.2.2) (2020-12-24)


### Bug Fixes

* update build script to properly detect OS ([b7b4699](https://www.github.com/socsieng/sendkeys-macos/commit/b7b46992fabb14511ac094939ce86ce405df4450))

### [1.2.1](https://www.github.com/socsieng/sendkeys-macos/compare/v1.2.0...v1.2.1) (2020-12-24)


### Bug Fixes

* include build scripts as part of package ([03b23ea](https://www.github.com/socsieng/sendkeys-macos/commit/03b23ea280e2857f5d4367458e27b2370813adf4))

## [1.2.0](https://www.github.com/socsieng/sendkeys-macos/compare/v1.1.2...v1.2.0) (2020-12-24)


### Features

* add experimental support for mouse movements ([7f5b93f](https://www.github.com/socsieng/sendkeys-macos/commit/7f5b93fe3cea210d587fe7b700064583b94aad2b))

### [1.1.2](https://www.github.com/socsieng/sendkeys-macos/compare/v1.1.1...v1.1.2) (2020-09-22)


### Bug Fixes

* exclude tests from npm package ([7b6f073](https://www.github.com/socsieng/sendkeys-macos/commit/7b6f073d7159d36016df75a1879c6e22f06c810c))

### [1.1.1](https://www.github.com/socsieng/sendkeys-macos/compare/v1.1.0...v1.1.1) (2020-09-22)


### Bug Fixes

* exclude images from npm package ([35bdb70](https://www.github.com/socsieng/sendkeys-macos/commit/35bdb70961af4ab0adeea82fd8d6c130f94a171b))

## [1.1.0](https://www.github.com/socsieng/sendkeys-macos/compare/v1.0.1...v1.1.0) (2020-09-18)


### Features

* add support for sticky pause ([133db1c](https://www.github.com/socsieng/sendkeys-macos/commit/133db1c53a22b7e34160acad73c2ab58ad1947f4))
* make application name optional ([9e519c4](https://www.github.com/socsieng/sendkeys-macos/commit/9e519c4300b1da03a6331ba93a21e6ed5e286cb2))

### [1.0.1](https://www.github.com/socsieng/sendkeys-macos/compare/v1.0.0...v1.0.1) (2020-09-17)


### Bug Fixes

* **build:** fix github workflow package name ([af97904](https://www.github.com/socsieng/sendkeys-macos/commit/af97904accfed7979b64532653b39c7466b8f7c7))

## [1.0.0](https://www.github.com/socsieng/sendkeys-macos/compare/v0.3.0...v1.0.0) (2020-09-17)


### ⚠ BREAKING CHANGES

* initial stable release

### Features

* initial stable release ([4f5fd95](https://www.github.com/socsieng/sendkeys-macos/commit/4f5fd9554fec8518a6a8802f73c37eb56782dcfa))


src/type.spec.js
const type = require('./type');
const sinon = require('sinon');
const expect = require('chai').expect;

function characterSequenceFromKeystrokes(keystrokeCalls) {
  return keystrokeCalls.map(call => call.args[0]).join('');
}

function getPauseSequenceFromDelays(delayCalls) {
  return delayCalls.map(call => call.args[0]);
}

describe('type', () => {
  const delay = sinon.spy();
  const keycode = sinon.spy();
  const keystroke = sinon.spy();

  const sysevents = {
    keyCode: keycode,
    keystroke,
  };

  beforeEach(() => {
    delay.resetHistory();
    keycode.resetHistory();
    keystroke.resetHistory();
  });

  it('should not fail when empty', () => {
    type('', 0.1, { delay, sysevents });

    expect(keystroke.called).to.equal(false);
    expect(delay.called).to.equal(false);
  });

  describe('keystrokes', () => {
    it('should send single keystroke', () => {
      type('1', 0.1, { delay, sysevents });

      expect(keystroke.calledWith('1')).to.equal(true);
      expect(delay.calledWith(0.1)).to.equal(true);
    });

    it('should send two keystrokes', () => {
      type('12', 0.1, { delay, sysevents });

      const sequence = characterSequenceFromKeystrokes(keystroke.getCalls());
      expect(sequence).to.equal('12');
      expect(delay.callCount).to.equal(2);
    });

    it('should send carriage returns instead of new lines', () => {
      type('1\n2\n', 0.1, { delay, sysevents });

      const sequence = characterSequenceFromKeystrokes(keystroke.getCalls());
      expect(sequence).to.equal('1\r2\r');
    });

    it('should send carriage returns instead of Windows style new lines', () => {
      type('1\r\n2\r\n', 0.1, { delay, sysevents });

      const sequence = characterSequenceFromKeystrokes(keystroke.getCalls());
      expect(sequence).to.equal('1\r2\r');
    });
  });

  describe('keycodes', () => {
    it('should send up keycode', () => {
      type('<c:up>', 0.1, { delay, sysevents });

      expect(keycode.calledWith(126, { using: [] })).to.equal(true);
    });

    it('should send up keycode with command down', () => {
      type('<c:up:command>', 0.1, { delay, sysevents });

      expect(keycode.calledWith(126, { using: ['command down'] })).to.equal(true);
    });

    it('should send up keycode with command down alias ⌘', () => {
      type('<c:up:⌘>', 0.1, { delay, sysevents });

      expect(keycode.calledWith(126, { using: ['command down'] })).to.equal(true);
    });

    it('should send up keycode with command down alias cmd', () => {
      type('<c:up:cmd>', 0.1, { delay, sysevents });

      expect(keycode.calledWith(126, { using: ['command down'] })).to.equal(true);
    });

    it('should send up keycode with multiple modifier keys', () => {
      type('<c:up:command,control,option,shift>', 0.1, { delay, sysevents });

      expect(
        keycode.calledWith(126, { using: ['command down', 'control down', 'option down', 'shift down'] }),
      ).to.equal(true);
    });

    it('should send up keycode with multiple modifier keys ignoring extra commas', () => {
      type('<c:up:,command,control,option,,>', 0.1, { delay, sysevents });

      expect(keycode.calledWith(126, { using: ['command down', 'control down', 'option down'] })).to.equal(true);
    });

    it('should not send keycodes when character sequence is upper case (C)', () => {
      type('<C:up:command,control,option,shift>', 0.1, { delay, sysevents });

      expect(keycode.notCalled).to.equal(true);
    });

    it('should send standard keystroke (2) with control down', () => {
      type('<c:2:control>', 0.1, { delay, sysevents });

      expect(keystroke.calledWith('2', { using: ['control down'] })).to.equal(true);
    });
  });

  describe('delays', () => {
    it('should pause for 1 second', () => {
      type('<p:1>', 0.1, { delay, sysevents });

      const delayCalls = delay.getCalls();
      expect(delayCalls[0].args).to.eql([1]);
    });

    it('should pause for 1.5 seconds', () => {
      type('<p:1.5>', 0.1, { delay, sysevents });

      const delayCalls = delay.getCalls();
      expect(delayCalls[0].args).to.eql([1.5]);
    });

    it('should pause for 1 and 1.5 seconds', () => {
      type('<p:1><p:1.5>', 0.1, { delay, sysevents });

      const delayCalls = delay.getCalls();
      expect(delayCalls[0].args).to.eql([1]);
      expect(delayCalls[1].args).to.eql([1.5]);
    });

    it('should pause for 1 and 1.5 seconds with character in between', () => {
      type('<p:1>a<p:1.5>', 0.1, { delay, sysevents });

      const delayCalls = delay.getCalls();
      expect(delayCalls[0].args).to.eql([1]);
      expect(delayCalls[1].args).to.eql([1.5]);
    });

    it('should pause for 1 second and revert to default', () => {
      type('<p:1>a', 0.1, { delay, sysevents });

      const delayCalls = delay.getCalls();
      expect(delayCalls[0].args).to.eql([1]);
      expect(delayCalls[1].args).to.eql([0.1]);
    });

    it('should remember sticky pause', () => {
      type('<P:1>abc', 0.1, { delay, sysevents });

      const keystrokeSequence = characterSequenceFromKeystrokes(keystroke.getCalls());
      const delaySequence = getPauseSequenceFromDelays(delay.getCalls());

      expect(keystrokeSequence).to.equal('abc');
      expect(delaySequence).to.eql([1, 1, 1, 1]);
    });

    it('should remember sticky pause mid sequence', () => {
      type('ab<P:1>cd', 0.1, { delay, sysevents });

      const keystrokeSequence = characterSequenceFromKeystrokes(keystroke.getCalls());
      const delaySequence = getPauseSequenceFromDelays(delay.getCalls());

      expect(keystrokeSequence).to.equal('abcd');
      expect(delaySequence).to.eql([0.1, 1, 1, 1]);
    });

    it('should remember sticky pause mid sequence with additional change in pause', () => {
      type('ab<P:1>cd<p:0.2>ef<P:3>ghi', 0.1, { delay, sysevents });

      const keystrokeSequence = characterSequenceFromKeystrokes(keystroke.getCalls());
      const delaySequence = getPauseSequenceFromDelays(delay.getCalls());

      expect(keystrokeSequence).to.equal('abcdefghi');
      expect(delaySequence).to.eql([0.1, 1, 1, 0.2, 1, 3, 3, 3, 3]);
    });
  });

  describe('continuations', () => {
    it('should do nothing when only character is a continuation', () => {
      type('<\\>', 0.1, { delay, sysevents });

      expect(keystroke.called).to.equal(false);
      expect(delay.called).to.equal(false);
    });

    it('should ignore the next character after continuation', () => {
      type('<\\>a', 0.1, { delay, sysevents });

      expect(keystroke.called).to.equal(false);
      expect(delay.called).to.equal(false);
    });

    it('should ignore the next character after continuation, but continue with next', () => {
      type('<\\>ab', 0.1, { delay, sysevents });

      expect(keystroke.calledWith('b')).to.equal(true);
      expect(delay.calledWith(0.1)).to.equal(true);
    });

    it('should ignore the first character in pause sequence', () => {
      type('<\\><p:1>', 0.1, { delay, sysevents });

      const sequence = characterSequenceFromKeystrokes(keystroke.getCalls());

      expect(sequence).to.equal('p:1>');
    });

    it('should ignore the first character in keycode sequence', () => {
      type('<\\><c:up:shift>', 0.1, { delay, sysevents });

      const sequence = characterSequenceFromKeystrokes(keystroke.getCalls());

      expect(sequence).to.equal('c:up:shift>');
    });
  });
});


src/splitter.spec.js
const splitter = require('./splitter');
const expect = require('chai').expect;

describe('splitter', () => {
  describe('single expression', () => {
    it('should split single expression', () => {
      const result = splitter.splitIncluding('hello<space>world', [/\<space\>/g]);

      expect(result).to.be.eql(['hello', '<space>', 'world']);
    });

    it('should split consecutive single expression', () => {
      const result = splitter.splitIncluding('hello<space><space>world', [/\<space\>/g]);

      expect(result).to.be.eql(['hello', '<space>', '<space>', 'world']);
    });

    it('should split single expression at start', () => {
      const result = splitter.splitIncluding('<space>world', [/\<space\>/g]);

      expect(result).to.be.eql(['<space>', 'world']);
    });

    it('should split single expression at end', () => {
      const result = splitter.splitIncluding('world<space>', [/\<space\>/g]);

      expect(result).to.be.eql(['world', '<space>']);
    });

    it('should split single expression near end', () => {
      const result = splitter.splitIncluding('world<space>a', [/\<space\>/g]);

      expect(result).to.be.eql(['world', '<space>', 'a']);
    });

    it('should throw for non global expressions', () => {
      expect(() => splitter.splitIncluding('<space>world', [/\<space\>/])).to.throw('Expression must be global');
    });
  });

  describe('multiple expressions', () => {
    it('should split multiple expression', () => {
      const result = splitter.splitIncluding('hello<space><tab>world', [/\<space\>/g, /\<tab\>/g]);

      expect(result).to.be.eql(['hello', '<space>', '<tab>', 'world']);
    });

    it('should split multiple expression at beginning and end', () => {
      const result = splitter.splitIncluding('<space>hello<tab>world<tab>', [/\<space\>/g, /\<tab\>/g]);

      expect(result).to.be.eql(['<space>', 'hello', '<tab>', 'world', '<tab>']);
    });
  });
});


src/mouse.m
#include <time.h>
#include <errno.h>    
#include <inttypes.h>
#include <math.h>

#include <Foundation/Foundation.h>
#include <ApplicationServices/ApplicationServices.h>

int msleep(long msec)
{
    struct timespec ts;
    int res;

    if (msec < 0)
    {
      errno = EINVAL;
      return -1;
    }

    ts.tv_sec = msec / 1000;
    ts.tv_nsec = (msec % 1000) * 1000000;

    do {
      res = nanosleep(&ts, &ts);
    } while (res && errno == EINTR);

    return res;
}

void setMousePosition(int x, int y, CGEventType state, CGMouseButton button) {
  CGEventRef move = CGEventCreateMouseEvent(
    NULL, state, // kCGEventMouseMoved,
    CGPointMake(x, y),
    button
  );

  CGEventPost(kCGHIDEventTap, move);

  CFRelease(move);
}

CGMouseButton getMouseButton(NSString *clickType) {
  CGMouseButton button = kCGMouseButtonLeft;

  if ([clickType isEqual:@"center"]) {
    button = kCGMouseButtonCenter;
  } else if ([clickType isEqual:@"right"]) {
    button = kCGMouseButtonRight;
  }

  return button;
}

void click(int x, int y, NSString *clickType, int clickCount) {
  CGEventType downEvent = kCGEventLeftMouseDown;
  CGEventType upEvent = kCGEventLeftMouseUp;
  CGMouseButton button = getMouseButton(clickType);

  if ([clickType isEqual:@"center"] || [clickType isEqual:@"right"]) {
    downEvent = kCGEventOtherMouseDown;
    upEvent = kCGEventOtherMouseUp;
  }

  CGEventRef down = CGEventCreateMouseEvent(
    NULL, downEvent,
    CGPointMake(x, y),
    button
  );
  CGEventSetIntegerValueField(down, kCGMouseEventClickState, clickCount);

  CGEventRef up = CGEventCreateMouseEvent(
    NULL, upEvent,
    CGPointMake(x, y),
    button
  );

  CGEventPost(kCGHIDEventTap, down);
  CGEventPost(kCGHIDEventTap, up);

  CFRelease(down);
  CFRelease(up);
}

float easeInOutQuad(float x) {
  return x < 0.5 ? 2 * x * x : 1 - pow(-2 * x + 2, 2) / 2;
}

int easePosition(int min, int max, float progress) {
  return ((max - min) * easeInOutQuad(progress)) + min;
}

long getCurrentMilliseconds() {
  struct timespec currentTime;
  clock_gettime(CLOCK_REALTIME, &currentTime);

  return round(currentTime.tv_nsec / 1.0e6) + (currentTime.tv_sec * 1000);
}

void move(int x1, int y1, int x2, int y2, int duration, NSString *drag) {
  struct timespec ts;
  int res;
  long startMilliseconds = getCurrentMilliseconds();
  long currentMilliseconds;

  CGEventType moveType = kCGEventMouseMoved;
  CGEventType downEvent = kCGEventLeftMouseDown;
  CGEventType upEvent = kCGEventLeftMouseUp;
  CGMouseButton button = getMouseButton(drag);
  CGEventRef down;
  CGEventRef up;

  // start drag
  if (drag != nil) {
    moveType = kCGEventOtherMouseDragged;

    if ([drag isEqual:@"center"] || [drag isEqual:@"right"]) {
      moveType = kCGEventLeftMouseDragged;
      downEvent = kCGEventOtherMouseDown;
      upEvent = kCGEventOtherMouseUp;
    }

    down = CGEventCreateMouseEvent(
      NULL, downEvent,
      CGPointMake(x1, y1),
      button
    );
    up = CGEventCreateMouseEvent(
      NULL, upEvent,
      CGPointMake(x2, y2),
      button
    );

    CGEventPost(kCGHIDEventTap, down);
  }

  ts.tv_sec = duration / 1000;
  ts.tv_nsec = (duration % 1000) * 1000000;

  setMousePosition(x1, y1, moveType, button);

  do {
    msleep(1);

    currentMilliseconds = getCurrentMilliseconds();
    float progress = (float)(currentMilliseconds - startMilliseconds) / (float)duration;
    if (progress > 1) {
      progress = 1;
    }

    int x = easePosition(x1, x2, progress);
    int y = easePosition(y1, y2, progress);

    setMousePosition(x, y, moveType, button);
  } while (currentMilliseconds - startMilliseconds <= duration);

  setMousePosition(x2, y2, moveType, button);

  // stop drag
  if (drag != nil) {
    CGEventPost(kCGHIDEventTap, up);

    CFRelease(down);
    CFRelease(up);
  }
}

CGPoint getCurrentMousePosition() {
  CGEventRef event = CGEventCreate(nil);
  CGPoint mouseLocation = CGEventGetLocation(event);
  CFRelease(event);

  return mouseLocation;
}

int main(int argc, char *argv[]) {
  NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
  NSUserDefaults *args = [NSUserDefaults standardUserDefaults];

  int x1 = [args integerForKey:@"x1"];
  int y1 = [args integerForKey:@"y1"];
  int x2 = [args integerForKey:@"x2"];
  int y2 = [args integerForKey:@"y2"];
  int duration = [args integerForKey:@"duration"];
  NSString *drag = [args stringForKey:@"drag"];
  NSString *clickType = [args stringForKey:@"click"];
  int clickCount = [args integerForKey:@"clicks"];
  int clickDelay = [args integerForKey:@"clickDelay"];

  CGPoint mouseLocation = getCurrentMousePosition();

  if (argc == 1) {
    printf("current position: %d, %d\n", (int)mouseLocation.x, (int)mouseLocation.y);
    return 0;
  }

  // use current mouse position
  if ([args objectForKey:@"x1"] == nil) {
    x1 = mouseLocation.x;
  }
  if ([args objectForKey:@"y1"] == nil) {
    y1 = mouseLocation.y;
  }

  if ([args objectForKey:@"x2"] == nil) {
    x2 = x1;
  }
  if ([args objectForKey:@"y2"] == nil) {
    y2 = y1;
  }

  move(x1, y1, x2, y2, duration, drag);

  if (clickCount == 0 && clickType != nil) {
    clickCount = 1;
  }

  if (clickCount > 0) {
    msleep(clickDelay);
    click(x2, y2, clickType, clickCount);
  }

  [pool release];
  return 0;
}


src/index.js
const { execFileSync } = require('child_process');
const generator = require('./generator');
const { splitIncluding } = require('./splitter');
const path = require('path');
const { existsSync } = require('fs');

const defaultOptions = {
  delay: 0,
  initialDelay: 0,
  generateScript: false,
};

// matches <m:1,2,3,4:0.5> or <m:1,2,3,4>
const mouseMoveExpression = /\<m:((\d+),(\d+),)?(\d+),(\d+)(:([\d.]+))?\>/;
// matches <m:left:1> or <m:right>
const mouseClickExpression = /\<m:([a-z]+)(:(\d+))?\>/;
// matches <d:1,2,3,4:0.5:left> or <d:1,2:0.5:left> or <d:1,2,3,4:0.5>
const mouseDragExpression = /\<d:((\d+),(\d+),)?(\d+),(\d+)(:([\d.]+))?(:([a-z]+))?\>/;

const mouseExpressions = [mouseMoveExpression, mouseClickExpression, mouseDragExpression];

function sendKeys(appName, /** @type {string} */ keystrokes, options = {}) {
  const mergedOptions = {
    ...defaultOptions,
    ...(options || {}),
  };

  if (mergedOptions.generateScript) {
    const output = generator.generate(keystrokes, appName, mergedOptions.delay, mergedOptions.initialDelay);
    process.stdout.write(output);
    return;
  }

  const usingMouse = hasMouseCommands(keystrokes);
  const hasMouse = existsSync(path.resolve(__dirname, '../build/mouse'));

  if (usingMouse && hasMouse) {
    // mouse support
    const fragments = splitIncluding(
      keystrokes,
      mouseExpressions.map(exp => new RegExp(exp, 'g')),
    );
    let isFirstActivation = true;

    for (const fragment of fragments) {
      let result;
      let match;

      if ((match = mouseMoveExpression.exec(fragment))) {
        const [, , x1, y1, x2, y2, , duration] = match;
        result = sendMouseMove(x1, y1, x2, y2, duration);
      } else if ((match = mouseClickExpression.exec(fragment))) {
        const [, button, , count] = match;
        result = sendMouseClick(button, count);
      } else if ((match = mouseDragExpression.exec(fragment))) {
        const [, , x1, y1, x2, y2, , duration, , button] = match;
        result = sendMouseDrag(x1, y1, x2, y2, duration, button);
      } else {
        const output = generator.generate(
          fragment,
          isFirstActivation ? appName : undefined,
          mergedOptions.delay,
          isFirstActivation ? mergedOptions.initialDelay : 0,
        );
        result = execFileSync('osascript', ['-l', 'JavaScript', '-e', output], {
          encoding: 'utf8',
          stdio: 'pipe',
        });
        isFirstActivation = false;
      }

      if (result && result.trim()) {
        console.log(result);
      }
    }
  } else {
    if (usingMouse && !hasMouse) {
      console.warn(
        'Script appears to making use of mouse commands without supported binary. Mouse commands will appear as keystrokes.',
      );
    }

    const output = generator.generate(keystrokes, appName, mergedOptions.delay, mergedOptions.initialDelay);
    const result = execFileSync('osascript', ['-l', 'JavaScript', '-e', output], {
      encoding: 'utf8',
      stdio: 'pipe',
    });

    if (result && result.trim()) {
      console.log(result);
    }
  }
}

function hasMouseCommands(keystrokes) {
  return mouseExpressions.map(exp => new RegExp(exp, 'g')).some(exp => exp.test(keystrokes));
}

function sendMouseMove(x1, y1, x2, y2, duration) {
  return sendMouseCommands({ x1, y1, x2, y2, duration });
}

function sendMouseClick(button, count) {
  return sendMouseCommands({ click: button, clicks: count });
}

function sendMouseDrag(x1, y1, x2, y2, duration, button) {
  return sendMouseCommands({ x1, y1, x2, y2, duration, drag: button || 'left' });
}

function sendMouseCommands(options) {
  const args = [];
  const fixedOptions = {
    ...options,
    duration: options.duration ? Number(options.duration) * 1000 : undefined,
  };

  Object.keys(fixedOptions)
    .filter(key => fixedOptions[key] != null)
    .forEach(key => args.push(`-${key}`, `${fixedOptions[key]}`));

  return execFileSync(path.resolve(__dirname, '../build/mouse'), args, {
    encoding: 'utf8',
    stdio: 'pipe',
  });
}

module.exports = sendKeys;


src/splitter.js
function splitIncluding(/** @type {string} */ input, /** @type {RegExp[]} */ expressions) {
  let result = [input];

  for (const expression of expressions) {
    const splitResults = result.map(item => splitAndIncludeExpression(item, expression));
    result = flatten(splitResults);
  }

  return result;
}

function splitAndIncludeExpression(/** @type {string} */ input, /** @type {RegExp} */ expression) {
  let match;
  const result = [];
  let index = 0;

  if (!expression.global) {
    throw new Error('Expression must be global');
  }

  while ((match = expression.exec(input))) {
    if (index !== match.index) {
      result.push(input.substring(index, match.index));
    }

    result.push(match[0]);
    index = match.index + match[0].length;
  }

  if (index < input.length) {
    result.push(input.substring(index));
  }

  return result;
}

function flatten(arr) {
  return Array.prototype.concat.apply([], arr);
}

module.exports = {
  splitIncluding,
};


src/type.js
/**
 * Types text input with a default delay between each character.
 *
 * @param {string} input
 * @param {number} defaultDelay
 * @param {*} options
 */
function type(input, defaultDelay, { delay, sysevents }) {
  const KEYS = {
    f1: 122,
    f2: 120,
    f3: 99,
    f4: 118,
    f5: 96,
    f6: 97,
    f7: 98,
    f8: 100,
    f9: 101,
    f10: 109,
    f11: 103,
    f12: 111,
    esc: 53,
    return: 36,
    enter: 76,
    delete: 51,
    space: 49,
    tab: 48,
    up: 126,
    down: 125,
    left: 123,
    right: 124,
    home: 115,
    end: 119,
    pgup: 116,
    pgdown: 121,
  };
  const MODIFIERS = {
    '⌘': 'command down',
    '^': 'control down',
    '⌥': 'option down',
    '⇧': 'shift down',
    'command': 'command down',
    'control': 'control down',
    'option': 'option down',
    'shift': 'shift down',
    'cmd': 'command down',
    'ctrl': 'control down',
    'alt': 'option down',
  };

  // matches <p:0.1>
  const pauseExpression = /^\<(p):([\d.]+)\>/i;
  // matches <c:down>, <c:up:command>
  const charExpression = /^\<c:(.|[\w]+)(:([,\w⌘^⌥⇧]+))?\>/;
  // matches <\>
  const continuationExpression = /^\<\\\>/;
  const lineEndingExpression = /\r?\n/g;

  const text = input.replace(lineEndingExpression, '\r');
  let index = 0;

  function ensurePause(pause) {
    const pauseMatch = pauseExpression.exec(text.substring(index));
    if (pauseMatch) {
      const value = parseFloat(pauseMatch[2]);
      delay(value);
      index += pauseMatch[0].length;

      // update default delay
      if (pauseMatch[1] === 'P') {
        defaultDelay = value;
      }

      return true;
    }
    if (pause) {
      delay(pause);
    }
    return false;
  }

  do {
    const continuationMatch = continuationExpression.exec(text.substring(index));

    if (continuationMatch) {
      index += continuationMatch[0].length + 1;
      continue;
    }

    const charMatch = charExpression.exec(text.substring(index));

    if (charMatch) {
      const value = charMatch[1];
      const modifiers = (charMatch[3] || '')
        .split(',')
        .filter(m => m)
        .map(m => MODIFIERS[m]);

      if (value in KEYS) {
        sysevents.keyCode(KEYS[value], { using: modifiers });
      } else {
        sysevents.keystroke(value[0], { using: modifiers });
      }
      index += charMatch[0].length;
      ensurePause(defaultDelay);
      continue;
    }

    const char = text[index];

    if (char) {
      if (ensurePause()) continue;

      sysevents.keystroke(char);
      index++;

      ensurePause(defaultDelay);
    }
  } while (text[index]);
}

module.exports = type;
